lib/SILOptimizer/SILCombiner/SILCombinerApplyVisitors.cpp

```diff
@@ -566,27 +566,62 @@ SILCombiner::optimizeConcatenationOfStringLiterals(ApplyInst *AI) {
   return tryToConcatenateStrings(AI, Builder);
 }
 
+// TODO:
+// Move this concrete type propagation for existentials out of sil-combine
+// We can then use DominanceAnalysis which is the right thing to do here
+// We avoid using it here due to compilation speed implications
+static bool fastDominates(SILInstruction *a, SILInstruction *b) {
+  auto aBlock = a->getParent(), bBlock = b->getParent();
+
+  // If the blocks are different, bail
+  if (aBlock != bBlock)
+    return true;
+
+  // Otherwise, they're in the same block, and we just need to check
+  // whether B comes after A.  This is a non-strict computation.
+  auto aIter = a->getIterator();
+  auto bIter = b->getIterator();
+  auto fIter = aBlock->begin();
+  while (bIter != fIter) {
+    --bIter;
+    if (aIter == bIter)
+      return true;
+  }
+
+  return false;
+}
+
 /// Returns the address of an object with which the stack location \p ASI is
 /// initialized. This is either a init_existential_addr or the source of a
 /// copy_addr. Returns a null value if the address does not dominate the
 /// alloc_stack user \p ASIUser.
 /// If the value is copied from another stack location, \p isCopied is set to
 /// true.
 static SILValue getAddressOfStackInit(AllocStackInst *ASI,
-                                      SILInstruction *ASIUser,
-                                      bool &isCopied) {
+                                      SILInstruction *ASIUser, bool &isCopied,
+                                      SILInstruction *OrigASIUser = nullptr) {
```

Comment:
Where is getAddressOfStackInit with OrigASIUser != null called?

Also, this new parameter should be documented.

Reply:
Used from
```
// Try to derive the type from the copy_addr that was used to
    // initialize the alloc_stack.
    assert(isCopied && "isCopied not set for a copy_addr");
    SILValue CAISrc = CAI->getSrc();
    if (auto *ASI = dyn_cast<AllocStackInst>(CAISrc))
      return getAddressOfStackInit(ASI, CAI, isCopied, ASIUser);
    return CAISrc;
```
Will add documentation